R/rtss.R

```diff
@@ -1,101 +1,182 @@
+  }
+}
+
+#' @title RTS,S mass vaccination listener
+#'
+#' @description schedules individuals to be vaccinated according to the mass
+#' strategy
+#'
+#' @param variables list of variables in the model
+#' @param events a list of events in the model
+#' @param parameters the model parameters
+#' @param correlations correlation parameters
+#' @noRd
+create_rtss_mass_listener <- function(
+  variables,
+  events,
+  parameters,
+  correlations
+  ) {
+  function(timestep) {
+    in_age_group <- individual::Bitset$new(parameters$human_population)
+    for (i in seq_along(parameters$rtss_mass_min_ages)) {
+      min_birth <- timestep - parameters$rtss_mass_max_ages[[i]]
+      max_birth <- timestep - parameters$rtss_mass_min_ages[[i]]
+      in_age_group$or(variables$birth$get_index_of(a = min_birth, b = max_birth))
+    }
+    if (parameters$rtss_mass_min_wait == 0) {
+      target <- in_age_group$to_vector()
+    } else {
+      not_recently_vaccinated <- variables$rtss_vaccinated$get_index_of(
+        a = max(timestep - parameters$rtss_mass_min_wait, 0),
+        b = timestep
+      )$not()
+      target <- in_age_group$and(not_recently_vaccinated)$to_vector()
     }
-    target <- which(target_indices & not_vaccinated)
+    
+    time_index = which(parameters$rtss_mass_timesteps == timestep)
     target <- target[
       sample_intervention(
         target,
-        'rtss',
-        parameters$rtss_coverages[[time_index]],
+       'rtss',
+        parameters$rtss_mass_coverages[[time_index]],
         correlations
       )
     ]
-    renderer$render('n_vaccinated', length(target), timestep)
-    if (length(target) > 0) {
-      variables$rtss_vaccinated$queue_update(
-        timestep,
-        target
+    schedule_vaccination(
+      target,
+      events,
+      parameters,
+      events$rtss_mass_doses
+    )
+    if (time_index < length(parameters$rtss_mass_timesteps)) {
+      events$rtss_mass_vaccination$schedule(
+        parameters$rtss_mass_timesteps[[time_index + 1]] - timestep
       )
     }
-    if (time_index < length(parameters$rtss_timesteps)) {
-      events$rtss_vaccination$schedule(
-        parameters$rtss_timesteps[[time_index + 1]] - timestep
-      )
+  }
+}
+
+#' @title Schedule vaccination doses and efficacy
+#'
+#' @param target vector of individuals to target
+#' @param variables list of variables in the model
+#' @param events a list of events in the model
+#' @param parameters the model parameters
+#' @param dose_events a list of dose events to schedule
+#' @noRd
+schedule_vaccination <- function(
+  target,
+  events,
+  parameters,
+  dose_events
+  ) {
+  if (length(target) > 0) {
+    for (d in seq_along(parameters$rtss_doses)) {
+      dose_events[[d]]$schedule(target, parameters$rtss_doses[[d]])
     }
-    if (length(parameters$rtss_boosters) > 0) {
-      events$rtss_booster$schedule(
-        target[bernoulli(length(target), parameters$rtss_booster_coverage[[1]])],
-        parameters$rtss_boosters[[1]]
```

Comment:
Let's check if the 'boosters' parameter in set_mass_rtss() is still timesteps following the initial vaccination. I'm wondering if it is now timesteps following the last dose. I can't locate the precise location where this would be in the code. 
![image](https://user-images.githubusercontent.com/80535782/128989125-741f1ded-54fd-4a4c-b827-b0530c24ae50.png)
^ seems like the blue line should be directly under the green

Reply:
It's after the last dose. I don't think I've changed that

Reply:
The code for scheduling the first booster is here: https://github.com/mrc-ide/malariasimulation/blob/feat/rtss_epi/R/rtss.R#L232-L254